tests/cxx/badinc.cc

```diff
@@ -213,7 +211,10 @@ Cc_tpl_typedef cc_tpl_typedef;
 // IWYU: I2_Class::InlFileTemplateFn is...*badinc-i2-inl.h
 // IWYU: I2_Class::InlFileStaticFn is...*badinc-i2-inl.h
 typedef I2_Class Cc_I2_Class_Typedef;
-// IWYU: I1_Struct needs a declaration
+// I1_Struct isn't really used by any possible operation with H_TemplateStruct,
```

Comment:
Sorry to come back to this at such a late stage, but I stumbled on the following bug report:
https://github.com/include-what-you-use/include-what-you-use/issues/1249#issuecomment-1541818734

Doesn't that sound kind of like the same behavior? In that case it's `shared_ptr<>` that triggers a requirement of its user-provided type for libstdc++ but not libc++. Probably their implementations are different in a way that triggers this. But it sounds like your change here would make that more the rule?

Would the workaround be to forward-declare the user type in the file instantiating `shared_ptr<>`?

Reply:
No, it is slightly different problem. `std::shared_ptr<T>::element_type` is defined through `std::remove_extent<T>::type` alias in libstdc++ and through `std::remove_extent_t<T>` alias template in libc++, and IWYU doesn't fully support alias templates at now. I performed [some work](https://github.com/bolshakov-a/include-what-you-use/commit/5e896ac6912bb735c43c10d1a202f81bc965605d) in that direction already, hoping it will arrive the main repository some day.